### DISCUSSION TITLE
iperf3: reverse mode json output

### DISCUSSION THREAD

**Discussion by jmauro** [esnet/iperf#1412] [Q&A]
Hello,

I have the following setup:
- a node as server:
```
$ iperf3 --version
iperf 3.9 (cJSON 1.7.13)
Linux XXXXX 5.15.0-1022-aws #26-Ubuntu SMP Thu Oct 13 12:59:25 UTC 2022 x86_64
Optional features available: CPU affinity setting, IPv6 flow label, SCTP, TCP congestion algorithm setting, sendfile / zerocopy, socket pacing, authentication
$ iperf3 --server --file /var/tmp/myfile --port 5005 --daemon --logfile /var/log/iperf3.log
```
- a client with the following setup:
```
$ iperf3 --port 5005 --reverse --parallel 10 --client X.X.X.X --json
```

This setup should be able to provide information regarding the possible download bandwidth of my client (do we agree on this?).

The issue, I am facing, is the `JSON`output:
```
                "sum_sent":     {
                        "start":        0,
                        "end":  10.036022,
                        "seconds":      10.036022,
                        "bytes":        5842668856,
                        "bits_per_second":      4657358348.5568285,
                        "retransmits":  58251,
                        "sender":       false
                },
                "sum_received": {
                        "start":        0,
                        "end":  10.000097,
                        "seconds":      10.000097,
                        "bytes":        5825119096,
                        "bits_per_second":      4660050074.31428,
                        "sender":       false
                },
                "cpu_utilization_percent":      {
                        "host_total":   71.8545657284829,
                        "host_user":    9.68909933186251,
                        "host_system":  62.165456479628567,
                        "remote_total": 35.028692900400273,
                        "remote_user":  0.99479011258444239,
                        "remote_system":        34.033869087365517
                },
                "sender_tcp_congestion":        "cubic",
                "receiver_tcp_congestion":      "cubic"
        }
```
I don't understand why there are such bit values on the `sum_sent` on the client JSON file, could someone enlighten me?

Even after reading the following [link](https://github.com/esnet/iperf/issues/480#issuecomment-307205313)

Kind regards,
JM

**Comment by davidBar-On** [comment]
Hello, I am not sure what is the problem you see. "bytes" of "sum_sent" show that about 5.8G bytes = 46G bits (5.8 x 8) were sent during 10 seconds.  I.e. about 4.6Gbits/sec as shown by the "bits_per_second".

**Reply by orientlu** [reply]
is L3 or L2

**Comment by jmauro** [comment]
The issue, I see : since I am in reverse mode there should be not a lot of sent traffic, or I am wrong ?

**Comment by davidBar-On** [comment]
Although the client reports the results, they are for the data sent from the server.  You can see that by the `"sender":       false` attribute.  In the non-reverse case it will be `"sender":       true`.

**Reply by jmauro** [reply]
Thanks for the tips.
I may have missed something, where is the documentation explaining this? Any pointer would be appreciated.

One more question, if I follow you, some figures are from the client and the other from the server which one is which?


Kind regards,
JM

**Comment by davidBar-On** [comment]
> I may have missed something, where is the documentation explaining this? Any pointer would be appreciated.

As far as I am aware, there is no documentation that describes the JSON output.  Maybe an issue should be opened about that.

> One more question, if I follow you, some figures are from the client and the other from the server which one is which?

The server is sending its statistics to the client and this is how the client is able to show both sides statistics.  The `"sender"` attribute indicates whether the client is the sender or not (there is no attribute that shows whether the statistics are from the client or from the server).

Note that in the server's json output the `"sender"` attribute indicates whether the server is the sender, and the statistics from the client are empty as the client does not send its statistics to the server.

When using using `--bidir` you can see that for the uplink the client is the sender and for the downlink it is not.

**Reply by jmauro** [reply]
Hello @davidBar-On 

Thanks for taking the time to answer.

I have created the [issue ](https://github.com/esnet/iperf/issues/1413) asking to update the documentation.

Kind regards,
JM
